server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -544,7 +552,15 @@ public ShardStartedClusterStateTaskExecutor(AllocationService allocationService,
             List<ShardRouting> shardRoutingsToBeApplied = new ArrayList<>(tasks.size());
             Set<ShardRouting> seenShardRoutings = new HashSet<>(); // to prevent duplicates
             for (StartedShardEntry task : tasks) {
-                ShardRouting matched = currentState.getRoutingTable().getByAllocationId(task.shardId, task.allocationId);
+                final IndexMetaData indexMetaData = currentState.metaData().index(task.shardId.getIndex());
```

Comment:
let's do this lookup only in `if (matched.primary() && task.primaryTerm > 0) {`, i.e., where it's actually needed.
Note that in that case the index metadata will be trivially non-null (assert indexMetaData != null), as there is a routing table for the shard.